go/ephemeral/ephemeral.go

```diff
@@ -188,13 +193,236 @@ func GetOwnDeviceEKs(ctx context.Context, g *libkb.GlobalContext) ([]keybase1.De
+	}
+
+	parsedResponse := UserEKResponse{}
+	err = res.Body.UnmarshalAgain(&parsedResponse)
+	if err != nil {
+		return nil, err
+	}
+
+	// The client now needs to verify several things about the blob its
+	// received:
+	// 1) Each is validly signed.
+	// 2) The signing key is the user's latest PUK.
+	// 3) The merkle payload supplied by the server matches the hash that's
+	//    been signed over.
+	// 4) The key hasn't expired. That is, the Merkle root it was delegated
+	//    with is within one week of the current root. The server deliberately
+	//    avoids doing this filtering for us, and finding expired keys in the
+	//    results here is expected. We silently drop them.
+
+	// Verify the sig.
+	signerKey, payload, _, err := libkb.NaclVerifyAndExtract(parsedResponse.Result.Sig)
+	if err != nil {
+		return nil, err
+	}
+
+	// Verify the signing key corresponds to the latest PUK. We load the user's
+	// UPAK from cache, but if the KID doesn't match, we try a forced reload to
+	// see if the cache might've been stale. Only if the KID still doesn't
+	// match after the reload do we complain.
+	upak, _, err := g.GetUPAKLoader().LoadV2(libkb.NewLoadUserByUIDArg(ctx, g, uid))
+	if err != nil {
+		return nil, err
+	}
+	latestPUK := upak.Current.GetLatestPerUserKey()
+	if latestPUK == nil || !latestPUK.SigKID.Equal(signerKey.GetKID()) {
+		// The latest PUK might be stale. Force a reload, then check this over again.
+		upak, _, err = g.GetUPAKLoader().LoadV2(libkb.NewLoadUserByUIDForceArg(g, uid))
+		if err != nil {
+			return nil, err
+		}
+		latestPUK = upak.Current.GetLatestPerUserKey()
+		if latestPUK == nil || !latestPUK.SigKID.Equal(signerKey.GetKID()) {
+			// It still looks stale after a fresh poll. Bail out.
+			latestPUKSigningKIDString := "<nil>"
+			if latestPUK != nil {
+				latestPUKSigningKIDString = fmt.Sprint(latestPUK.SigKID)
+			}
+			return nil, fmt.Errorf("userEK returned for UID %s PUK signing KID %s, but latest is %s",
+				uid, signerKey.GetKID(), latestPUKSigningKIDString)
+		}
+	}
+
+	// Decode the signed JSON.
+	var verifiedMetadata keybase1.UserEkMetadata
+	err = json.Unmarshal(payload, &verifiedMetadata)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check the hash of the Merkle payload, confirm it matches what was signed, and parse it.
+	computedHash := sha256.Sum256([]byte(parsedResponse.Result.MerklePayload))
+	if !bytes.Equal(verifiedMetadata.HashMeta, computedHash[:]) {
+		return nil, fmt.Errorf("supplied merkle root doesn't match signed hash meta")
+	}
+	var signedMerkleRoot libkb.MerkleRootPayloadUnpacked
+	err = json.Unmarshal([]byte(parsedResponse.Result.MerklePayload), &signedMerkleRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check whether the key is expired. This isn't considered an error,
+	// since the server doesn't do this check for us. We log these cases
+	// and return nil.
+	currentMerkleRoot, err := g.GetMerkleClient().FetchRootFromServer(ctx, libkb.EphemeralKeyMerkleFreshness)
```

Comment:
we should split the above to it's own helper function since we'll need to get a generation from the server even if the key is expired (if our cache is empty) or we missed some generation and the sever rejects us

Reply:
Could you clarify this one?

Reply:
resolved offline